File: server-code/spring/epa-air/src/main/java/org/openapitools/model/Air05.java
package org.openapitools.model;

import java.net.URI;
import java.util.Objects;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonTypeName;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.openapitools.model.Air03;
import org.openapitools.model.Air04;
import org.openapitools.jackson.nullable.JsonNullable;
import java.time.OffsetDateTime;
import javax.validation.Valid;
import javax.validation.constraints.*;
import io.swagger.v3.oas.annotations.media.Schema;


import java.util.*;
import javax.annotation.Generated;

/**
 * Results Object
 */

@Schema(name = "air05", description = "Results Object")
@JsonTypeName("air05")
@Generated(value = "org.openapitools.codegen.languages.SpringCodegen", date = "2024-01-06T13:50:47.778831-04:00[America/Lower_Princes]")
public class Air05 {

  private String badSystemIDs;

  private String cvRows;

  private String feARows;

  @Valid
  private List<@Valid Air03> facilities;

  private String inSPRows;

  private String indianCountryRows;

  private String infFEARows;

  private Air04 mapOutput;

  private String message;

  private String pageNo;

  private String queryID;

  private String queryRows;

  private String svRows;

  private String totalPenalties;

  private String v3Rows;

  public Air05() {
    super();
  }

  /**
   * Constructor with only required parameters
   */
  public Air05(String badSystemIDs, String cvRows, String feARows, String inSPRows, String indianCountryRows, String infFEARows, String message, String pageNo, String queryID, String queryRows, String svRows, String totalPenalties, String v3Rows) {
    this.badSystemIDs = badSystemIDs;
    this.cvRows = cvRows;
    this.feARows = feARows;
    this.inSPRows = inSPRows;
    this.indianCountryRows = indianCountryRows;
    this.infFEARows = infFEARows;
    this.message = message;
    this.pageNo = pageNo;
    this.queryID = queryID;
    this.queryRows = queryRows;
    this.svRows = svRows;
    this.totalPenalties = totalPenalties;
    this.v3Rows = v3Rows;
  }

  public Air05 badSystemIDs(String badSystemIDs) {
    this.badSystemIDs = badSystemIDs;
    return this;
  }

  /**
   * Identifies which passed query system identifiers are invalid.
   * @return badSystemIDs
  */
  @NotNull 
  @Schema(name = "BadSystemIDs", description = "Identifies which passed query system identifiers are invalid.", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("BadSystemIDs")
  public String getBadSystemIDs() {
    return badSystemIDs;
  }

  public void setBadSystemIDs(String badSystemIDs) {
    this.badSystemIDs = badSystemIDs;
  }

  public Air05 cvRows(String cvRows) {
    this.cvRows = cvRows;
    return this;
  }

  /**
   * Summary count of the number of CWA facilities or SDWA public drinking water systems with current violations.
   * @return cvRows
  */
  @NotNull 
  @Schema(name = "CVRows", example = "0", description = "Summary count of the number of CWA facilities or SDWA public drinking water systems with current violations.", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("CVRows")
  public String getCvRows() {
    return cvRows;
  }

  public void setCvRows(String cvRows) {
    this.cvRows = cvRows;
  }

  public Air05 feARows(String feARows) {
    this.feARows = feARows;
    return this;
  }

  /**
   * Summary count of the number of facilities with a formal enforcement action in the past five years
   * @return feARows
  */
  @NotNull 
  @Schema(name = "FEARows", example = "1", description = "Summary count of the number of facilities with a formal enforcement action in the past five years", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("FEARows")
  public String getFeARows() {
    return feARows;
  }

  public void setFeARows(String feARows) {
    this.feARows = feARows;
  }

  public Air05 facilities(List<@Valid Air03> facilities) {
    this.facilities = facilities;
    return this;
  }

  public Air05 addFacilitiesItem(Air03 facilitiesItem) {
    if (this.facilities == null) {
      this.facilities = new ArrayList<>();
    }
    this.facilities.add(facilitiesItem);
    return this;
  }

  /**
   * A complex array of facility information.
   * @return facilities
  */
  @Valid 
  @Schema(name = "Facilities", description = "A complex array of facility information.", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
  @JsonProperty("Facilities")
  public List<@Valid Air03> getFacilities() {
    return facilities;
  }

  public void setFacilities(List<@Valid Air03> facilities) {
    this.facilities = facilities;
  }

  public Air05 inSPRows(String inSPRows) {
    this.inSPRows = inSPRows;
    return this;
  }

  /**
   * Number of facilities with insp_5yr_flag populated (CWP_DATE_LAST_INSPECTION)
   * @return inSPRows
  */
  @NotNull 
  @Schema(name = "INSPRows", example = "1", description = "Number of facilities with insp_5yr_flag populated (CWP_DATE_LAST_INSPECTION)", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("INSPRows")
  public String getInSPRows() {
    return inSPRows;
  }

  public void setInSPRows(String inSPRows) {
    this.inSPRows = inSPRows;
  }

  public Air05 indianCountryRows(String indianCountryRows) {
    this.indianCountryRows = indianCountryRows;
    return this;
  }

  /**
   * Number of facilities with tribal_flag populated
   * @return indianCountryRows
  */
  @NotNull 
  @Schema(name = "IndianCountryRows", example = "0", description = "Number of facilities with tribal_flag populated", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("IndianCountryRows")
  public String getIndianCountryRows() {
    return indianCountryRows;
  }

  public void setIndianCountryRows(String indianCountryRows) {
    this.indianCountryRows = indianCountryRows;
  }

  public Air05 infFEARows(String infFEARows) {
    this.infFEARows = infFEARows;
    return this;
  }

  /**
   * Number of facilities with infea_5yr_flag populated (INFORMAL_ENF_ACT_COUNT > 0)
   * @return infFEARows
  */
  @NotNull 
  @Schema(name = "InfFEARows", example = "0", description = "Number of facilities with infea_5yr_flag populated (INFORMAL_ENF_ACT_COUNT > 0)", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("InfFEARows")
  public String getInfFEARows() {
    return infFEARows;
  }

  public void setInfFEARows(String infFEARows) {
    this.infFEARows = infFEARows;
  }

  public Air05 mapOutput(Air04 mapOutput) {
    this.mapOutput = mapOutput;
    return this;
  }

  /**
   * Get mapOutput
   * @return mapOutput
  */
  @Valid 
  @Schema(name = "MapOutput", requiredMode = Schema.RequiredMode.NOT_REQUIRED)
  @JsonProperty("MapOutput")
  public Air04 getMapOutput() {
    return mapOutput;
  }

  public void setMapOutput(Air04 mapOutput) {
    this.mapOutput = mapOutput;
  }

  public Air05 message(String message) {
    this.message = message;
    return this;
  }

  /**
   * Field to record messages (typically performance-related) about packet processing
   * @return message
  */
  @NotNull 
  @Schema(name = "Message", example = "Success", description = "Field to record messages (typically performance-related) about packet processing", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("Message")
  public String getMessage() {
    return message;
  }

  public void setMessage(String message) {
    this.message = message;
  }

  public Air05 pageNo(String pageNo) {
    this.pageNo = pageNo;
    return this;
  }

  /**
   * The number of pages of results returned
   * @return pageNo
  */
  @NotNull 
  @Schema(name = "PageNo", example = "1", description = "The number of pages of results returned", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("PageNo")
  public String getPageNo() {
    return pageNo;
  }

  public void setPageNo(String pageNo) {
    this.pageNo = pageNo;
  }

  public Air05 queryID(String queryID) {
    this.queryID = queryID;
    return this;
  }

  /**
   * Sequential number assigned to entire search result
   * @return queryID
  */
  @NotNull 
  @Schema(name = "QueryID", example = "245", description = "Sequential number assigned to entire search result", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("QueryID")
  public String getQueryID() {
    return queryID;
  }

  public void setQueryID(String queryID) {
    this.queryID = queryID;
  }

  public Air05 queryRows(String queryRows) {
    this.queryRows = queryRows;
    return this;
  }

  /**
   * Number of query results returned
   * @return queryRows
  */
  @NotNull 
  @Schema(name = "QueryRows", example = "5", description = "Number of query results returned", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("QueryRows")
  public String getQueryRows() {
    return queryRows;
  }

  public void setQueryRows(String queryRows) {
    this.queryRows = queryRows;
  }

  public Air05 svRows(String svRows) {
    this.svRows = svRows;
    return this;
  }

  /**
   * Number of facilities with curr_sv_flag populated (CWP_STATUS = \"Significant Violation\")
   * @return svRows
  */
  @NotNull 
  @Schema(name = "SVRows", example = "0", description = "Number of facilities with curr_sv_flag populated (CWP_STATUS = \"Significant Violation\")", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("SVRows")
  public String getSvRows() {
    return svRows;
  }

  public void setSvRows(String svRows) {
    this.svRows = svRows;
  }

  public Air05 totalPenalties(String totalPenalties) {
    this.totalPenalties = totalPenalties;
    return this;
  }

  /**
   * The total dollar amount of either assessed or final penalties within the five year time period
   * @return totalPenalties
  */
  @NotNull 
  @Schema(name = "TotalPenalties", example = "$963,531", description = "The total dollar amount of either assessed or final penalties within the five year time period", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("TotalPenalties")
  public String getTotalPenalties() {
    return totalPenalties;
  }

  public void setTotalPenalties(String totalPenalties) {
    this.totalPenalties = totalPenalties;
  }

  public Air05 v3Rows(String v3Rows) {
    this.v3Rows = v3Rows;
    return this;
  }

  /**
   * Number of facilities having one or more quarters in non-compliance (QNC) in the last three years
   * @return v3Rows
  */
  @NotNull 
  @Schema(name = "V3Rows", example = "2", description = "Number of facilities having one or more quarters in non-compliance (QNC) in the last three years", requiredMode = Schema.RequiredMode.REQUIRED)
  @JsonProperty("V3Rows")
  public String getV3Rows() {
    return v3Rows;
  }

  public void setV3Rows(String v3Rows) {
    this.v3Rows = v3Rows;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Air05 air05 = (Air05) o;
    return Objects.equals(this.badSystemIDs, air05.badSystemIDs) &&
        Objects.equals(this.cvRows, air05.cvRows) &&
        Objects.equals(this.feARows, air05.feARows) &&
        Objects.equals(this.facilities, air05.facilities) &&
        Objects.equals(this.inSPRows, air05.inSPRows) &&
        Objects.equals(this.indianCountryRows, air05.indianCountryRows) &&
        Objects.equals(this.infFEARows, air05.infFEARows) &&
        Objects.equals(this.mapOutput, air05.mapOutput) &&
        Objects.equals(this.message, air05.message) &&
        Objects.equals(this.pageNo, air05.pageNo) &&
        Objects.equals(this.queryID, air05.queryID) &&
        Objects.equals(this.queryRows, air05.queryRows) &&
        Objects.equals(this.svRows, air05.svRows) &&
        Objects.equals(this.totalPenalties, air05.totalPenalties) &&
        Objects.equals(this.v3Rows, air05.v3Rows);
  }

  @Override
  public int hashCode() {
    return Objects.hash(badSystemIDs, cvRows, feARows, facilities, inSPRows, indianCountryRows, infFEARows, mapOutput, message, pageNo, queryID, queryRows, svRows, totalPenalties, v3Rows);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class Air05 {\n");
    sb.append("    badSystemIDs: ").append(toIndentedString(badSystemIDs)).append("\n");
    sb.append("    cvRows: ").append(toIndentedString(cvRows)).append("\n");
    sb.append("    feARows: ").append(toIndentedString(feARows)).append("\n");
    sb.append("    facilities: ").append(toIndentedString(facilities)).append("\n");
    sb.append("    inSPRows: ").append(toIndentedString(inSPRows)).append("\n");
    sb.append("    indianCountryRows: ").append(toIndentedString(indianCountryRows)).append("\n");
    sb.append("    infFEARows: ").append(toIndentedString(infFEARows)).append("\n");
    sb.append("    mapOutput: ").append(toIndentedString(mapOutput)).append("\n");
    sb.append("    message: ").append(toIndentedString(message)).append("\n");
    sb.append("    pageNo: ").append(toIndentedString(pageNo)).append("\n");
    sb.append("    queryID: ").append(toIndentedString(queryID)).append("\n");
    sb.append("    queryRows: ").append(toIndentedString(queryRows)).append("\n");
    sb.append("    svRows: ").append(toIndentedString(svRows)).append("\n");
    sb.append("    totalPenalties: ").append(toIndentedString(totalPenalties)).append("\n");
    sb.append("    v3Rows: ").append(toIndentedString(v3Rows)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }
}

File: server-code/aspnetcore/epa-air/src/Org.OpenAPITools/Models/Air05.cs
/*
 * U.S. EPA Enforcement and Compliance History Online (ECHO) - Clean Air Act
 *
 * Enforcement and Compliance History Online (ECHO) is a tool developed and maintained by EPA's Office of Enforcement and Compliance Assurance for public use. ECHO provides integrated compliance and enforcement information for over 1 million regulated facilities nationwide.  Air Rest Services provides multiple service endpoints, each with specific capabilities, to search and retrieve data on facilities regulated under the Clean Air Act (CAA). The returned results reflect data drawn from EPA's ICIS-Air database.  The get_facilities, get_map, get_qid, and get_download end points are meant to be used together, while the enhanced get_facility_info end point is self contained. The get_facility_info end point returns either an array of state, county or zip clusters with summary statistics per cluster or an array of facilities.  The recommended use scenario for get_facilities, get_qid, get_map, and get_downoad is:  <b>1)</b>  Use get_facilities to validate passed query parameters, obtain summary statistics and to obtain a query_id (QID).  QIDs are time sensitive and will be valid for approximately 30 minutes. <b>2)</b>  Use get_qid, with the returned QID, to paginate through arrays of facility results. <b>3)</b>  Use get_map, with the returned QID, to zoom in/out and pan on the clustered and individual facility coordinates that meet the QID query criteria. <b>4)</b>  Use get_download, with the returned QID, to generate a Comma Separated Value (CSV) file of facility information that meets the QID query criteria.  Use the qcolumns parameter to customize your search results.  Use the Metadata service endpoint for a list of available output objects, their Column Ids, and their definitions to help you build your customized output.   Additional ECHO Resources:   <a href=\"https://echo.epa.gov/tools/web-services\">Web Services</a>, <a href=\"https://echo.epa.gov/resources/echo-data/about-the-data\">About ECHO's Data</a>, <a href=\"https://echo.epa.gov/tools/data-downloads\">Data Downloads</a>
 *
 * The version of the OpenAPI document: 2019.10.15
 * 
 * Generated by: https://openapi-generator.tech
 */

using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Org.OpenAPITools.Converters;

namespace Org.OpenAPITools.Models
{ 
    /// <summary>
    /// Results Object
    /// </summary>
    [DataContract]
    public partial class Air05 : IEquatable<Air05>
    {
        /// <summary>
        /// Identifies which passed query system identifiers are invalid.
        /// </summary>
        /// <value>Identifies which passed query system identifiers are invalid.</value>
        [Required]
        [DataMember(Name="BadSystemIDs", EmitDefaultValue=false)]
        public string BadSystemIDs { get; set; }

        /// <summary>
        /// Summary count of the number of CWA facilities or SDWA public drinking water systems with current violations.
        /// </summary>
        /// <value>Summary count of the number of CWA facilities or SDWA public drinking water systems with current violations.</value>
        /// <example>0</example>
        [Required]
        [DataMember(Name="CVRows", EmitDefaultValue=false)]
        public string CVRows { get; set; }

        /// <summary>
        /// Summary count of the number of facilities with a formal enforcement action in the past five years
        /// </summary>
        /// <value>Summary count of the number of facilities with a formal enforcement action in the past five years</value>
        /// <example>1</example>
        [Required]
        [DataMember(Name="FEARows", EmitDefaultValue=false)]
        public string FEARows { get; set; }

        /// <summary>
        /// A complex array of facility information.
        /// </summary>
        /// <value>A complex array of facility information.</value>
        [DataMember(Name="Facilities", EmitDefaultValue=false)]
        public List<Air03> Facilities { get; set; }

        /// <summary>
        /// Number of facilities with insp_5yr_flag populated (CWP_DATE_LAST_INSPECTION)
        /// </summary>
        /// <value>Number of facilities with insp_5yr_flag populated (CWP_DATE_LAST_INSPECTION)</value>
        /// <example>1</example>
        [Required]
        [DataMember(Name="INSPRows", EmitDefaultValue=false)]
        public string INSPRows { get; set; }

        /// <summary>
        /// Number of facilities with tribal_flag populated
        /// </summary>
        /// <value>Number of facilities with tribal_flag populated</value>
        /// <example>0</example>
        [Required]
        [DataMember(Name="IndianCountryRows", EmitDefaultValue=false)]
        public string IndianCountryRows { get; set; }

        /// <summary>
        /// Number of facilities with infea_5yr_flag populated (INFORMAL_ENF_ACT_COUNT &gt; 0)
        /// </summary>
        /// <value>Number of facilities with infea_5yr_flag populated (INFORMAL_ENF_ACT_COUNT &gt; 0)</value>
        /// <example>0</example>
        [Required]
        [DataMember(Name="InfFEARows", EmitDefaultValue=false)]
        public string InfFEARows { get; set; }

        /// <summary>
        /// Gets or Sets MapOutput
        /// </summary>
        [DataMember(Name="MapOutput", EmitDefaultValue=false)]
        public Air04 MapOutput { get; set; }

        /// <summary>
        /// Field to record messages (typically performance-related) about packet processing
        /// </summary>
        /// <value>Field to record messages (typically performance-related) about packet processing</value>
        /// <example>Success</example>
        [Required]
        [DataMember(Name="Message", EmitDefaultValue=false)]
        public string Message { get; set; }

        /// <summary>
        /// The number of pages of results returned
        /// </summary>
        /// <value>The number of pages of results returned</value>
        /// <example>1</example>
        [Required]
        [DataMember(Name="PageNo", EmitDefaultValue=false)]
        public string PageNo { get; set; }

        /// <summary>
        /// Sequential number assigned to entire search result
        /// </summary>
        /// <value>Sequential number assigned to entire search result</value>
        /// <example>245</example>
        [Required]
        [DataMember(Name="QueryID", EmitDefaultValue=false)]
        public string QueryID { get; set; }

        /// <summary>
        /// Number of query results returned
        /// </summary>
        /// <value>Number of query results returned</value>
        /// <example>5</example>
        [Required]
        [DataMember(Name="QueryRows", EmitDefaultValue=false)]
        public string QueryRows { get; set; }

        /// <summary>
        /// Number of facilities with curr_sv_flag populated (CWP_STATUS &#x3D; \&quot;Significant Violation\&quot;)
        /// </summary>
        /// <value>Number of facilities with curr_sv_flag populated (CWP_STATUS &#x3D; \&quot;Significant Violation\&quot;)</value>
        /// <example>0</example>
        [Required]
        [DataMember(Name="SVRows", EmitDefaultValue=false)]
        public string SVRows { get; set; }

        /// <summary>
        /// The total dollar amount of either assessed or final penalties within the five year time period
        /// </summary>
        /// <value>The total dollar amount of either assessed or final penalties within the five year time period</value>
        /// <example>$963,531</example>
        [Required]
        [DataMember(Name="TotalPenalties", EmitDefaultValue=false)]
        public string TotalPenalties { get; set; }

        /// <summary>
        /// Number of facilities having one or more quarters in non-compliance (QNC) in the last three years
        /// </summary>
        /// <value>Number of facilities having one or more quarters in non-compliance (QNC) in the last three years</value>
        /// <example>2</example>
        [Required]
        [DataMember(Name="V3Rows", EmitDefaultValue=false)]
        public string V3Rows { get; set; }

        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.Append("class Air05 {\n");
            sb.Append("  BadSystemIDs: ").Append(BadSystemIDs).Append("\n");
            sb.Append("  CVRows: ").Append(CVRows).Append("\n");
            sb.Append("  FEARows: ").Append(FEARows).Append("\n");
            sb.Append("  Facilities: ").Append(Facilities).Append("\n");
            sb.Append("  INSPRows: ").Append(INSPRows).Append("\n");
            sb.Append("  IndianCountryRows: ").Append(IndianCountryRows).Append("\n");
            sb.Append("  InfFEARows: ").Append(InfFEARows).Append("\n");
            sb.Append("  MapOutput: ").Append(MapOutput).Append("\n");
            sb.Append("  Message: ").Append(Message).Append("\n");
            sb.Append("  PageNo: ").Append(PageNo).Append("\n");
            sb.Append("  QueryID: ").Append(QueryID).Append("\n");
            sb.Append("  QueryRows: ").Append(QueryRows).Append("\n");
            sb.Append("  SVRows: ").Append(SVRows).Append("\n");
            sb.Append("  TotalPenalties: ").Append(TotalPenalties).Append("\n");
            sb.Append("  V3Rows: ").Append(V3Rows).Append("\n");
            sb.Append("}\n");
            return sb.ToString();
        }

        /// <summary>
        /// Returns the JSON string presentation of the object
        /// </summary>
        /// <returns>JSON string presentation of the object</returns>
        public string ToJson()
        {
            return JsonConvert.SerializeObject(this, Formatting.Indented);
        }

        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="obj">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object obj)
        {
            if (obj is null) return false;
            if (ReferenceEquals(this, obj)) return true;
            return obj.GetType() == GetType() && Equals((Air05)obj);
        }

        /// <summary>
        /// Returns true if Air05 instances are equal
        /// </summary>
        /// <param name="other">Instance of Air05 to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(Air05 other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;

            return 
                (
                    BadSystemIDs == other.BadSystemIDs ||
                    BadSystemIDs != null &&
                    BadSystemIDs.Equals(other.BadSystemIDs)
                ) && 
                (
                    CVRows == other.CVRows ||
                    CVRows != null &&
                    CVRows.Equals(other.CVRows)
                ) && 
                (
                    FEARows == other.FEARows ||
                    FEARows != null &&
                    FEARows.Equals(other.FEARows)
                ) && 
                (
                    Facilities == other.Facilities ||
                    Facilities != null &&
                    other.Facilities != null &&
                    Facilities.SequenceEqual(other.Facilities)
                ) && 
                (
                    INSPRows == other.INSPRows ||
                    INSPRows != null &&
                    INSPRows.Equals(other.INSPRows)
                ) && 
                (
                    IndianCountryRows == other.IndianCountryRows ||
                    IndianCountryRows != null &&
                    IndianCountryRows.Equals(other.IndianCountryRows)
                ) && 
                (
                    InfFEARows == other.InfFEARows ||
                    InfFEARows != null &&
                    InfFEARows.Equals(other.InfFEARows)
                ) && 
                (
                    MapOutput == other.MapOutput ||
                    MapOutput != null &&
                    MapOutput.Equals(other.MapOutput)
                ) && 
                (
                    Message == other.Message ||
                    Message != null &&
                    Message.Equals(other.Message)
                ) && 
                (
                    PageNo == other.PageNo ||
                    PageNo != null &&
                    PageNo.Equals(other.PageNo)
                ) && 
                (
                    QueryID == other.QueryID ||
                    QueryID != null &&
                    QueryID.Equals(other.QueryID)
                ) && 
                (
                    QueryRows == other.QueryRows ||
                    QueryRows != null &&
                    QueryRows.Equals(other.QueryRows)
                ) && 
                (
                    SVRows == other.SVRows ||
                    SVRows != null &&
                    SVRows.Equals(other.SVRows)
                ) && 
                (
                    TotalPenalties == other.TotalPenalties ||
                    TotalPenalties != null &&
                    TotalPenalties.Equals(other.TotalPenalties)
                ) && 
                (
                    V3Rows == other.V3Rows ||
                    V3Rows != null &&
                    V3Rows.Equals(other.V3Rows)
                );
        }

        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            unchecked // Overflow is fine, just wrap
            {
                var hashCode = 41;
                // Suitable nullity checks etc, of course :)
                    if (BadSystemIDs != null)
                    hashCode = hashCode * 59 + BadSystemIDs.GetHashCode();
                    if (CVRows != null)
                    hashCode = hashCode * 59 + CVRows.GetHashCode();
                    if (FEARows != null)
                    hashCode = hashCode * 59 + FEARows.GetHashCode();
                    if (Facilities != null)
                    hashCode = hashCode * 59 + Facilities.GetHashCode();
                    if (INSPRows != null)
                    hashCode = hashCode * 59 + INSPRows.GetHashCode();
                    if (IndianCountryRows != null)
                    hashCode = hashCode * 59 + IndianCountryRows.GetHashCode();
                    if (InfFEARows != null)
                    hashCode = hashCode * 59 + InfFEARows.GetHashCode();
                    if (MapOutput != null)
                    hashCode = hashCode * 59 + MapOutput.GetHashCode();
                    if (Message != null)
                    hashCode = hashCode * 59 + Message.GetHashCode();
                    if (PageNo != null)
                    hashCode = hashCode * 59 + PageNo.GetHashCode();
                    if (QueryID != null)
                    hashCode = hashCode * 59 + QueryID.GetHashCode();
                    if (QueryRows != null)
                    hashCode = hashCode * 59 + QueryRows.GetHashCode();
                    if (SVRows != null)
                    hashCode = hashCode * 59 + SVRows.GetHashCode();
                    if (TotalPenalties != null)
                    hashCode = hashCode * 59 + TotalPenalties.GetHashCode();
                    if (V3Rows != null)
                    hashCode = hashCode * 59 + V3Rows.GetHashCode();
                return hashCode;
            }
        }

        #region Operators
        #pragma warning disable 1591

        public static bool operator ==(Air05 left, Air05 right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(Air05 left, Air05 right)
        {
            return !Equals(left, right);
        }

        #pragma warning restore 1591
        #endregion Operators
    }
}
